src/compiler/checker.ts

```diff
@@ -32018,6 +32018,41 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         );
     }
 
+    function discriminateContextualTypeByArrayElements(node: ArrayLiteralExpression, contextualType: UnionType) {

```

Comment:
`length` can be a special kind of discriminator here, if element size of the array literal expression is known (no spreads, or spreads with fixed length) then the compiler should be able to reject types that have other numeric literal `length`s

Reply:
Do you mean I should discard a type like `{ 0: number, length: 10 }` because it's inconsistant? Could you be more specific?

Reply:
In a situation like this:
```ts
declare function test(_: [(arg: number) => void] | [(arg: string) => void, true]): void

test([arg => {
  arg // number
}])
```

we can safely ignore the second union constituent. It has `length: 2` and it's guaranteed that our argument must have `length: 1`.

Reply:
This is already working with my code:

```
if (elementsLength > properties.length || elementsLength < properties.reduce((c, p) => p.flags & SymbolFlags.Optional ? c : ++c, 0)) return false;
```

Reply:
`elementsLength > properties.length` doesn't look correct to me. I just tested out `getPropertiesOfType` on `[string, string]` and it returns 43 items because it includes all prototype methods 🤔

Reply:
This is fixed with my new commit